api/src/routes/challenge.ts

```diff
@@ -424,5 +429,95 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.get(
+    '/exam/:id',
+    {
+      schema: schemas.exam,
+      errorHandler(error, request, reply) {
+        if (error.validation) {
+          void reply.code(400);
+          return { error: `Valid 'id' not found in request parameters.` };
+        } else {
+          fastify.errorHandler(error, request, reply);
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        const { id } = req.params;
+
+        const { completedChallenges } =
+          await fastify.prisma.user.findUniqueOrThrow({
+            where: { id: req.session.user.id },
+            select: { completedChallenges: true }
+          });
+
+        const examFromDb = await fastify.prisma.exam.findUnique({
+          where: { id }
+        });
+
+        console.log(examFromDb);
+
+        if (!examFromDb) {
+          void reply.code(500);
+          return {
+            error: 'An error occurred trying to get the exam from the database.'
+          };
+        }
+
+        const validExamFromDbSchema = validateExamFromDbSchema(examFromDb);
+
+        if ('error' in validExamFromDbSchema) {
+          void reply.code(500);
+          return {
+            error:
+              'An error occurred validating the exam information from the database.'
+          };
+        }
+
+        const { prerequisites, numberOfQuestionsInExam, title } = examFromDb;
+
+        // Validate User has completed prerequisite challenges
+        for (let i = 0; i < prerequisites.length; i++) {
+          const prerequisiteCompleted = completedChallenges.some(
+            challenge => challenge.id === prerequisites[i]?.id
+          );
+
+          if (!prerequisiteCompleted) {
+            void reply.code(403);
+            return {
+              error: `You have not completed the required challenges to start the '${title}'.`
+            };
+          }
+        }
```

Comment:
We want to confirm all _prerequisite_ (subset) challenges are in `completedChallenges` (set).

Small improvement would be to loop over `completedChallenges` - once - and confirm every `prerequisite` is included. I probably would have done something like (untested):

```js
const prerequisiteIds = prerequisites.map(p => p.id);
const complete = completedChallenges.filter(c => {
    return prerequisiteIds.includes(c.id);
});

if (complete.length !== prerequisites.length) {
  return "Error"
}
```

It does not seem any better, thinking about it more 🤷‍♂️